docs/design-documents/20221028-dead-letter-queue.md

```diff
@@ -0,0 +1,192 @@
+  connector should be preserved, even if messages are nacked in a different
+  order (this is already assured using
+  a [semaphore](https://github.com/ConduitIO/conduit/blob/b6325584d51ec64bc9086faddcdaa2788a5dfa8f/pkg/pipeline/stream/source_acker.go#L35-L37)).
+- If a message failed to be sent to the DLQ the nack action should fail and
+  stop the pipeline.
+- Allow the user to configure the window and number of messages that need to be
+  nacked to consider the pipeline broken and trigger a stop. Example: if 25 of
+  the last 30 messages are nacked stop the pipeline.
+- Messages pushed to the DLQ should contain metadata describing the error and
+  the source connector.
+
+### Questions
+
+- **Should we allow the user to attach custom metadata to messages pushed to
+  the DLQ?**
+
+### Background
+
+Currently Conduit pipelines fail fast. If a message fails to be processed for
+any reason it is negatively acknowledged which causes the pipeline to stop
+running as soon as possible. In reality it's often the case that we don't want
+to stop the pipeline because of one bad message, instead we want to store it
+somewhere safe and keep the pipeline running. The solution is the support for
+dead letter queues or DLQs.
+
+## Implementation
+
+### DLQ Handlers
+
+Internally we already have a place to implement DLQ handlers
+([here](https://github.com/ConduitIO/conduit/blob/d19379efc04d20d12ab9c80df82a29fcef7e8afd/pkg/pipeline/stream/source_acker.go#L114-L118)).
+We need to decide on the API of the DLQ handler, in this design we propose an
+interface with a method that takes a context, a record and an error and returns
+an error. This should be all we need to write that record to the DLQ.
+
+```go
+type DLQHandler struct {
+    func Write(ctx context.Context, r record.Record, reason error) error
+}
+```
+
+The function `DLQHandler.Write` is expected to synchronously write the record
+to the DLQ and return `nil` if that action succeeded or an error otherwise. The
+synchronous write is needed to ensure the DLQ message actually reached its
+destination, otherwise we need to consider the nack failed and need to stop the
+pipeline.
+
+### DLQ Strategies
+
+We propose to implement 2 DLQ strategies:
+
+- `log` - This strategy would cause a nacked message to be logged to the
+  Conduit log. Unless a stop window and threshold are configured the pipeline
+  would keep running.
+- `plugin` - This strategy would reroute nacked messages to a destination
+  connector plugin. This strategy needs to be able to get additional
+  configuration options defining the plugin name and its settings.
+
+Regardless of the strategy the user can also configure a stop window and stop
+threshold.
+
+- `stop_window` defines how many last acks/nacks are monitored in the window
+  that controls if the pipeline should stop. 0 disables the window.
+- `stop_threshold` defines the number of nacks in the window that would cause
```

Comment:
A percentage would be good too (depending on the volume of messages, the same number of nacks may be relevant or irrelevant).

Reply:
I started out with a percentage but switched to an absolute number. My reasoning was that "90% of 30 messages" or "27 of 30 messages" are both the same, although the second one is easier to grasp. Also if the config says "92% of 30 messages" I'm not exactly sure what the threshold is, 27 or 28? So I went with the absolute number for clarity.

I hope that's what you meant, if not let me know.

Reply:
Btw - any thoughts about the naming? I'm not sure I'm _entirely_ happy with `stop_window` and `stop_threshold`, if anyone has better suggestions I'm all ears 🙂

Reply:
I think I should have expressed myself better in the first comment.:) The reason I suggested percentages too (in addition to what you already described for the window of last N messages) is to have a global limit (vs. local in what's in the doc).

It could be that due to a reason last N out of N messages got nacked, but because of many more ack'd messages before them, the overall success rate is very high and acceptable. That might have been a glitch and things continue to work as expected after the glitch.

On the other hand, it may be a sign of an issue which is worsening, so someone would like to focus on a sliding window.

In other words, I see two different options for a limit on the errors: global and local, and I see use cases for both.

> Btw - any thoughts about the naming? I'm not sure I'm _entirely_ happy with `stop_window` and `stop_threshold`, if anyone has better suggestions I'm all ears slightly_smiling_face

A slightly different alternative I can suggest is `sliding_window_size`/`sliding_window_threshold` or just `window_size`/`threshold` or `limit`. Adding the `sliding` prefix might  make it more clear. Since these come in the DLQ section, just saying `threshold`/`limit` is maybe enough (i.e. I think it can be easily understood that it's the DLQ threshold/limit and that it will cause a stop). I don't think it's really far from what you already suggested though.:)

Reply:
Let me see if I understand this - you're saying we should also introduce a global limit of nacked messages as a percentage of all processed messages? I see a few problems with this. Let's say we start a pipeline and the first message gets nacked, globally that's 100% of all messages, so that will necessarily stop the pipeline if any global limit is configured. On the other hand, if a pipeline is running for a _long_ time, the global limit will take a long time to take effect, even if all messages start getting nacked (after 1 year a 1% limit would present roughly 3,5 days worth of nacked messages). I think the usefulness of a global limit is _limited_ (pun intended 🙃). It only works as expected within a certain frame, that's what the window is trying to address.

> It could be that due to a reason last N out of N messages got nacked /.../ That might have been a glitch and things continue to work as expected after the glitch.

Isn't that the purpose of configuring a stop window? It lets you automatically stop the pipeline when it starts misbehaving. If such a "glitch" stops a pipeline without the user wanting that to happen, then they should configure a larger window and/or threshold, because they apparently don't care about N nacked messages in a row.

---

Regarding the variable naming, you're right maybe `stop` is redundant. I'd also say that `sliding` would be redundant, I don't imagine what other window this would be if not a sliding one 😄 What about the combination `window_size` and `window_nack_threshold`?

Reply:
> Isn't that the purpose of configuring a stop window?

I'd say it's for sure one of the purposes, but I imagine there are also cases where you don't care about some local peaks, as long as you're overall in some limits. However, I don't think at all it's a big thing with or without it, nor would the design radically change, so I'm fine leaving it out, especially for the sake of unblocking the design doc and progressing.

> What about the combination `window_size` and `window_nack_threshold`?

I like those!